internal/filesystem/ipfs/shared.go

```diff
@@ -392,3 +471,67 @@ func fsTypeName(mode fs.FileMode) string {
 		return "irregular"
 	}
 }
+
+func makeAndAddLink(ctx context.Context, target string, dag coreiface.APIDagService) (cid.Cid, error) {
+	dagData, err := unixfs.SymlinkData(target)
+	if err != nil {
+		return cid.Cid{}, err
+	}
+	dagNode := mdag.NodeWithData(dagData)
+	if err := dag.Add(ctx, dagNode); err != nil {
+		return cid.Cid{}, err
+	}
+	return dagNode.Cid(), nil
+}
+
+func getUnixFSLink(ctx context.Context,
+	op, name string,
+	ufs coreiface.UnixfsAPI, cid cid.Cid,
+	allowedPrefix string,
+) (string, error) {
+	cPath := corepath.IpfsPath(cid)
+	link, err := ufs.Get(ctx, cPath)
+	if err != nil {
+		const kind = fserrors.IO
+		return "", fserrors.New(op, name, err, kind)
+	}
+	return resolveNodeLink(op, name, link, allowedPrefix)
+}
+
+func resolveNodeLink(op, name string, node files.Node, prefix string) (string, error) {
+	target, err := readNodeLink(op, name, node)
+	if err != nil {
+		return "", err
+	}
+	// We allow 2 kinds of absolute links:
+	// 1) File system's root
+	// 2) Paths matching an explicitly allowed prefix
+	if strings.HasPrefix(target, prefix) {
+		target = strings.TrimPrefix(target, prefix)
+		return path.Clean(target), nil
+	}
+	switch target {
+	case "/":
+		return filesystem.Root, nil
+	case "..":
+		name = path.Dir(name)
+		fallthrough
+	case ".":
+		return path.Dir(name), nil
+	}
+	if target[0] == '/' {
+		const (
+			err  = generic.ConstError("link target must be relative")
+			kind = fserrors.InvalidItem
+		)
+		pair := fmt.Sprintf(
+			`%s -> %s`,
+			name, target,
+		)
+		return "", fserrors.New(op, pair, err, kind)
+	}
+	if target = path.Join("/"+name, target); target == "/" {
```

Comment:
does this work on windows?

Reply:
In this context it's not supposed to.
Go's `fs.FS` paths are non-rooted `/` separated paths.
And this call is only used in an `fs.FS` context.
When our system resolves links, it wants that path format.
The `ReadLink` method is an exception, which returns the raw symlink string no matter what it is.
But we don't use that method to traverse links internally.

That said, I do think there's a bug here.
I'm rooting the path at `/`, and joining it, in case there's any dots that climb all the way up to the `path` root (`/`), and if so converting it to the `fs.FS` root (`.`).
But I'm never taking that `/` back out, which creates an invalid `fs.FS` path.
I.e. `/go/fs/standard` should actually be `go/fs/standard`.

Reply:
Resolved the issue where I was returning a `/` rooted path (that `fs.FS` does not want)
https://github.com/djdv/go-filesystem-utils/compare/01526aa45558472b148a6ea0309a84ec0d84a0a7..841152d326564906a4a0db301ac10af143bff37d